ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala

```diff
@@ -271,15 +271,29 @@ private[kvutils] class ProcessTransactionSubmission(
         }
         ._1
 
-      r <- if (allUnique)
+      _ <- if (allUnique)
         pass
       else
         reject(
           recordTime,
           buildRejectionLogEntry(
             transactionEntry,
             RejectionReason.Disputed("DuplicateKey: Contract Key not unique")))
-    } yield r
+
+      // LookupByKey nodes themselves don't actually fetch the contract.
+      // Therefore we need to do an additional check on all contract keys
+      // that the referred contract satisfies the causal monotonicity invariant.
+      causalKeyMonotonicity = startingKeys.forall { key =>
```

Comment:
To improve readability, can you move this into a helper function, please? This function is already too long.

Reply:
Fixed in 14a595f.